zmq_actor.hpp

```diff
@@ -0,0 +1,120 @@
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#ifndef __ZMQ_ACTOR_HPP_INCLUDED__
+#define __ZMQ_ACTOR_HPP_INCLUDED__
+
+#include "zmq.hpp"
+
+#ifdef ZMQ_CPP11
+// Actor needs some threading.
+// Initial implementation uses std::thread so C++11 only.
+#include <thread>
+
+namespace zmq {
+
+    typedef std::pair<zmq::socket_t, zmq::socket_t> pipe;
+
+    /*! Return a pair of PAIR sockets.
+     *
+     * First is bound and second is connected via inproc://
+     */
+    pipe create_pipe(context_t& ctx)
+    {
+        pipe ret{socket_t(ctx, socket_type::pair),
+                 socket_t(ctx, socket_type::pair)};
+
+        std::stringstream ss;
+        ss << "inproc://pipe-"
+           << std::hex
+           << ret.first.handle()
+           << "-"
+           << ret.second.handle();
+        std::string addr = ss.str();
+        ret.first.bind(addr.c_str());
+        ret.second.connect(addr.c_str());
+        return ret;
+    }
+
+
+    // This shim calls the actor function and then provides hard-wired
+    // protocol for shutdown, mimicking how CZMQ zactor_t does it.
+    template<typename Func, typename... Args>
+    void actor_shim(socket_t&& apipe, Func fn, Args... args) {
+        fn(apipe, args...);
+        apipe.send(zmq::message_t{}, zmq::send_flags::none);
+    }
+
+    /*! An CZMQ zactor_t like class.
+     *
+     * Construct it with an actor function and any args.  
+     *
+     * The function will be spawned in a thread.
+     *
+     * Application may use pipe() for a socket with which to
+     * communicate with the actor.
+     *
+     * The destructor will signal through the pipe and wait for the
+     * actor function to exit, if indeed it is still around.
+     */
+    class actor_t {
+    public:
+
+        template<typename Func, typename... Args>
+        actor_t(context_t& ctx, Func fn, Args... args) {
+            socket_t apipe;
+            std::tie(apipe, _pipe) = create_pipe(ctx);
+                              
+            _thread = std::thread(actor_shim<Func, Args...>,
+                                  std::move(apipe),
+                                  fn,
+                                  args...);
+
+            zmq::message_t rmsg; // wait for actor ready signal
+            auto res = _pipe.recv(rmsg);
+        }
+
+        ~actor_t() {
+            auto sres = _pipe.send(message_t("$TERM",5), send_flags::dontwait);
+            if (sres) {
+                message_t rmsg;
+                auto res = _pipe.recv(rmsg, recv_flags::none);
```

Comment:
Is this required, is joining the thread not sufficient?

Reply:
This `recv()` implements the minimal actor shutdown protocol that CZMQ implements.

Reply:
I think this blocking recv should be omitted also. I also think the shutdown message signal should be a empty message, to be consistent with the thread shutdown signal, and also similar to closing a ZMQ_STREAM connection. That would simply be
```c++
try {
  _pipe.send(zmq::const_buffer{}, send_flags::dontwait);
} catch (...) {} // catch exception in case of context shutdown
_thread.join();
```

Reply:
Hi @gummif.  Changing this mini-protocol will mean CZMQ actor functions can not be used as cppzmq actor functions.  Maybe there is a way to make an optional adapter function to provide a compatibility layer.  If so, I see no objection to this and your other suggestion and they will make the base code simpler. I'll look into it.  Thanks.

Reply:
For this and the other comments regarding emulating CZMQ's actor startup/shutdown mini-protocol, I've come to accept the minimalist approach you suggest.  I think the application will need to invent something for practical purposes but that is maybe best left to the user.  The only downside is that the shutdown part is best put in an actor destructor. 

I did add an example of implementing this at the application level in the tests.

This then leaves `zmq::actor_t` as just providing a bundling of thread/func/link and providing the type erasure of the func.